tools/cli/commands/connect.py

```diff
@@ -0,0 +1,264 @@
+
+        After the instance is reachable, the `on_ready` method is called.
+
+        This method is meant to be suitable for running in a separate thread,
+        and takes an event argument to indicate when that thread should exit.
+
+        Args:
+          cancelled_event: A threading.Event instance that indicates we should
+            give up on the instance becoming reachable.
+        """
+        health_url = '{0}_info/'.format(datalab_address)
+        healthy = False
+        print('Waiting for Datalab to be reachable at ' + datalab_address)
+        while not cancelled_event.is_set():
+            try:
+                health_resp = urllib2.urlopen(health_url)
+                if health_resp.getcode() == 200:
+                    healthy = True
+                    break
+            except:
+                continue
+
+        if healthy:
+            on_ready()
+        return
+
+    def connect_and_check():
+        """Create a connection to Datalab and notify the user when ready.
+
+        This method blocks for as long as the connection is open.
+
+        Raises:
+          KeyboardInterrupt: If the user kills the connection.
+        """
+        cancelled_event = threading.Event()
+        health_check_thread = threading.Thread(
+            target=health_check,
+            args=[cancelled_event])
+        health_check_thread.start()
+        try:
+            create_tunnel()
+        except subprocess.CalledProcessError:
+            print('Connection broken')
+        finally:
+            cancelled_event.set()
+            health_check_thread.join()
+        return
+
+    remaining_reconnects = args.max_reconnects
+    while True:
+        try:
+            connect_and_check()
+        except KeyboardInterrupt:
+            return
+        if remaining_reconnects == 0:
+            return
+        print('Attempting to reconnect...')
+        remaining_reconnects -= 1
+        # Don't launch the browser on reconnect...
+        args.launch_browser = False
+    return
+
+
+def maybe_start(args, gcloud_compute, instance):
+    """Start the given Google Compute Engine VM if it is not running.
+
+    Args:
+      args: The Namespace instance returned by argparse
+      gcloud_compute: Function that can be used to invoke `gcloud compute`
+      instance: The name of the instance to check and (possibly) start
+    Raises:
+      subprocess.CalledProcessError: If one of the `gcloud` calls fails
+    """
+    get_cmd = ['instances', 'describe']
+    start_cmd = ['instances', 'start']
+    if args.zone:
+        get_cmd.extend(['--zone', args.zone])
+        start_cmd.extend(['--zone', args.zone])
+    get_cmd.extend(['--format', 'value(status)', instance])
+    start_cmd.extend([instance])
+    status = 'UNKNOWN'
+    with tempfile.TemporaryFile() as tf:
+        gcloud_compute(args, get_cmd, stdout=tf)
+        tf.seek(0)
+        status = tf.read().strip()
+    if status != 'RUNNING':
```

Comment:
What if a non-existent instance name is supplied here (e.g. by mistake)? Does this result in forever trying to connect?